main/src/library/Map.flix

```diff
@@ -1,7 +1,266 @@
 namespace Map {
 
-    def empty[k, v]: Map[k, v] = ???
+    //
+    // ## Map Construction
+    //
 
-    def insert[k, v](k: k, v: v, m: Map[k, v]): Map[k, v] = ???
+    ///
+    /// Returns the empty map.
+    ///
+    def empty[k, v]: Map[k, v] = Map(Nil)
 
+    ///
+    /// Returns the singleton map where key `k` is mapped to value `v`.
+    ///
+    def singleton[k, v](k: k, v: v): Map[k, v] = Map((k, v) :: Nil)
+
+    //
+    // ## Basic Operations
+    //
+
+    ///
+    /// Returns `true` if and only if `m` is the empty map, i.e. `Map(Nil)`.
+    ///
+    def null[k, v](m: Map[k, v]): Bool =
+        let Map(xs) = m;
+            List.isEmpty(xs)
+
+    ///
+    /// Returns `Some(v)` if `k -> v` is in `m`.
+    /// Otherwise returns `None`.
+    ///
+    def get[k, v](k: k, m: Map[k, v]): Option[v] = match m with {
+        case Map(Nil) => None
+        case Map((a, b) :: rs) => if (a == k) Some(b) else get(k, Map(rs))
+    }
+
+    ///
+    /// Returns `v` if `k -> v` is in `m`.
+    /// Otherwise, returns `d`.
+    ///
+    def getWithDefault[k, v](k: k, d: v, m: Map[k, v]): v = match m with {
+        case Map(Nil) => d
+        case Map((a, b) :: rs) => if (a == k) b else getWithDefault(k, d, Map(rs))
+    }
+
+    ///
+    /// Returns `true` if and only if `m` contains the key `k`.
+    ///
+    def memberOf[k, v](k: k, m: Map[k, v]): Bool = match m with {
+        case Map(Nil) => false
+        case Map((a, b) :: rs) => if (a == k) true else memberOf(k, Map(rs))
+    }
+
+    ///
+    /// Returns the keys of `m`.
+    ///
+    def keysOf[k, v](m: Map[k, v]): Set[k] =
+        let Map(xs) = m;
+            Set(fst(List.unzip(xs)))
+
+    ///
+    /// Returns the values of `m`.
+    ///
+    def valuesOf[k, v](m: Map[k, v]): List[v] =
+        let Map(xs) = m;
+            snd(List.unzip(xs))
+
+    //
+    // ## Insert
+    //
+
+    ///
+    /// Updates `m` with `k -> v`.
+    ///
+    def insert[k, v](k: k, v: v, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(insertHelper(k, v, xs))
```

Comment:
I'm thinking that `insert` (and all its variants) could be expressed in terms of `insertWithKey` (which is the most general version, right?)